spec/requests/graphql/mutations/create_asset_spec.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+require 'digest'
+
+RSpec.describe Mutations::CreateAsset, type: :request do
+  let(:authorized_object) { FactoryBot.create(:item, :with_project, :with_asset) }
+  let(:authorized_project) { authorized_object.projects.first }
+  let(:blob_content) { "This is text to store in a blob" }
+  let(:blob_checksum) { Digest::MD5.hexdigest blob_content }
+  let(:blob_args) do
+    {
+      filename: 'blob.xyz',
+      byte_size: blob_content.bytesize,
+      checksum: blob_checksum,
+      content_type: 'text/plain'
+    }
+  end
+  let(:active_storage_blob) do
+    blob = ActiveStorage::Blob.create_before_direct_upload!(**blob_args)
+    blob.upload(StringIO.new(blob_content))
+    blob
+  end
+
+  include_examples 'requires user to have correct permissions for graphql request' do
+    let(:variables) do
+      {
+        input: {
+          parentId: authorized_object.uid, signedBlobId: 'not-relevant'
+        }
+      }
+    end
+    let(:request) { graphql query, variables }
+  end
+
+  context 'when logged in user is an administrator' do
+    before { sign_in_user as: :administrator }
+
+    context 'when adding asset from an upload' do
+      let(:variables) do
+        {
+          input: {
+            parentId: authorized_object.uid, signedBlobId: active_storage_blob.signed_id
+          }
+        }
+      end
+
+      before { graphql query, variables }
+
+      it 'returns a new asset' do
+        expect(response.body).to have_json_type(String).at_path('data/createAsset/asset/id')
+      end
+      it 'deletes the upload' do
```

Comment:
Can you also add a test that confirms that the copied file doesn't exist if the asset fails to save (confirming that `cleanup_on_failure`) is being called and doing its job?

Reply:
This is harder to do with random resource storage locations.

Reply:
Maybe you can mock the uri generation response?  I'm just concerned that we're not testing it.

Reply:
Ok, an awkward test but it is in place.

Reply:
Thanks for adding that.  Checking it out now.